nx/lib/nx/binary_backend.ex

```diff
@@ -150,8 +150,17 @@ defmodule Nx.BinaryBackend do
   @impl true
   def from_binary(t, binary, _backend_options), do: from_binary(t, binary)
 
-  defp from_binary(t, binary) when is_binary(binary), do: %{t | data: %B{state: binary}}
-  defp from_binary(t, other), do: %{t | data: %B{state: IO.iodata_to_binary(other)}}
+  defp from_binary(%{type: {_, bitsize}, shape: shape} = t, binary) when is_binary(binary) do
+    expected_size = Tuple.product(shape) * div(bitsize, 8)
+
+    if byte_size(binary) != expected_size do
```

Comment:
I've added this to help catch some bugs. Should we turn this into a configurable or dev-only behavior?

Reply:
We check this in `Nx.from_backend`. What could be bypassing the check there? 🤔

Reply:
There are lot's of places that call this defp directly in BinaryBackend. Basically everything that goes into binary-world calls this in the end